### PR TITLE
[v8.0] Match all the things

### PR DESCRIPTION
**Pull request by chrisburr**
In LHCbDIRAC we found that the matcher was struggling when there are many jobs running at a single site. This was caused by the JobDB being slow to do `select Type, count(*) from Jobs where Site = %(site)s group by Type`. Every 10 seconds the cache would expire and every thread would try to query the DB again and hang for a long time.

Looking at the culmative response time over a couple of hours (with a similar number of requests in both periods) the new version is 98.9% faster.

Before this change:

Screenshot 2024-11-22 at 12 55 22

Screenshot 2024-11-22 at 14 20 42

After this change:

Screenshot 2024-11-22 at 12 55 40

Screenshot 2024-11-22 at 14 20 56

BEGINRELEASENOTES

*WorkloadManagment
CHANGE: Better caching performance in the Matching Limiter

ENDRELEASENOTES

### CODE REVIEW

File: src/DIRAC/WorkloadManagementSystem/Client/Limiter.py [hunk 1]
```diff
@@ -12,10 +21,109 @@
...
+            self.futures.pop(key)
+            self.hard_cache[key] = result
+            self.soft_cache[key] = result
+
+
 class Limiter:
     # static variables shared between all instances of this class
     csDictCache = DictCache()
     condCache = DictCache()
+    newCache = TwoLevelCache(10, 300)
```
Comment (chrisburr) [inline src/DIRAC/WorkloadManagementSystem/Client/Limiter.py:126]: It's probably worth replacing `condCache` with `newCache` but I don't have time for it.

File: src/DIRAC/WorkloadManagementSystem/Client/Limiter.py [hunk 2]
```diff
@@ -249,3 +345,15 @@ def __getDelayCondition(self, siteName):
...
+    def _countsByJobType(self, siteName, attName):
+        result = self.jobDB.getCounters(
+            "Jobs",
+            [attName],
+            {"Site": siteName, "Status": [JobStatus.RUNNING, JobStatus.MATCHED, JobStatus.STALLED]},
+        )
+        if not result["OK"]:
+            return result
+        data = result["Value"]
+        data = {k[0][attName]: k[1] for k in data}
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Client/Limiter.py:358]:
```suggestion
        data = {k[0][attName]: k[1] for k in result["Value"]}
```

File: src/DIRAC/WorkloadManagementSystem/Client/Limiter.py [hunk 3]
```diff
@@ -12,10 +21,109 @@
...
+            return result
+        with self.locks[key]:
+            if key not in self.futures:
+                self.futures[key] = self.pool.submit(self._work, key, populate_func)
+            if result := self.hard_cache.get(key):
+                self.soft_cache[key] = result
+                return result
+            # It is critical that ``future`` is waited for outside of the lock as
+            # _work aquires the lock before filling the caches. This also means
+            # we can gaurentee that the future has not yet been removed from the
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Client/Limiter.py:95]:
```suggestion
            # we can guarantee that the future has not yet been removed from the
```

File: src/DIRAC/WorkloadManagementSystem/Client/Limiter.py [hunk 4]
```diff
@@ -12,10 +21,109 @@
 from DIRAC.WorkloadManagementSystem.Client import JobStatus
 
 
+class TwoLevelCache:
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Client/Limiter.py:24]: It does not seem to me there's anything "specific" to this class. What if you move it to a generic utility module?

File: src/DIRAC/WorkloadManagementSystem/Client/Limiter.py [hunk 5]
```diff
@@ -12,10 +21,109 @@
...
+
+    def __init__(self, soft_ttl: int, hard_ttl: int, *, max_workers: int = 10, max_items: int = 1_000_000):
+        """Initialize the TwoLevelCache with specified TTLs."""
+        self.soft_cache = TTLCache(max_items, soft_ttl)
+        self.hard_cache = TTLCache(max_items, hard_ttl)
+        self.locks = defaultdict(threading.Lock)
+        self.futures: dict[str, Future] = {}
+        self.pool = ThreadPoolExecutor(max_workers=max_workers)
+
+    def get(self, key: str, populate_func: Callable[[], Any]):
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Client/Limiter.py:63]:
```suggestion
    def get(self, key: str, populate_func: Callable[[], Any]) -> dict:
```

**Review by fstagni — APPROVED** [review]

**Comment by fstagni** [comment]
I will merge this one right now, and take care of minor comments (and possibly other usage of the new cache) in a later PR.

**Pull request merged**